### ISSUE TITLE
oauth2 authentication fails because qgis sends a query parameter without a corresponding value

### ISSUE THREAD

**Issue by ricardogsilva** [qgis/QGIS#37925]
**Describe the bug**

I've got a local geonode that is setup with some layers. This geonode is serving as an oauth2 provider for geoserver, which in turn is publishing the layers via WMS. The layers are not public.

Geonode is running at `http://localhost/`

I want to use QGIS to access said layers via WMS and for that I need to setup a suitable oauth2 connection.

Things almost work OK, but QGIS seems to produce an invalid authorization request, when contacting GeoNode. The request being sent to GeoNode is (line breaks added by me):

```http
http://localhost/o/authorize/?
    response_type=code&
    client_id=TiwJNIXeFTDBK59DdOwMrHKbyPuOG6LWpPG53O1K&
    redirect_uri=http://127.0.0.1:7070/&
    scope=read%20write%20groups&
    state&
    api_key=
```

pay special attention to the **state** query parameter which is being added to the request. This parameter is specified as a single value, instead of the more popular `key=value` form. 

On my particular case, GeoNode cannot parse the query string due to this parameter. I get an error message in the browser saying:

```
Error: invalid_request
Unable to parse query string
```

A simple workaround for this is to simply replace `state` with `state=` in the original request. This makes the authentication flow succeed without problem and I am able to load the layers inside QGIS.

I don't know what the `state` parameter is supposed to mean, but this seems like a pretty easy fix.

**How to Reproduce**

Unfortunately, reproducing this requires quite a lot of setup. You'll need to setup a GeoNode instance with a proper oauth2 provider configuration. The general steps are:

1. Clone geonode repo and use [geonode's docker-compose] file to standup a geonode stack (geonode, geoserver, postgis db)

1. Login to geonode and access its admin area (default credentials are `user: admin` `password: admin`)

1. In the admin, go to _DJANGO OAUTH TOOLKIT -> Applications_. Create a new application with the following settings:

    -  name should reflect the client and the user. for example `admin qgis`
    -  user should be the user you will want to use in QGIS (e.g. admin)
    -  set a redirect uri of `http://127.0.0.1:7070` -> the string must match what QGIS will request, and by default this is it
    -  take note of the automatically-generated `client id` and `client secret` values, as these are used by QGIS
    -  set `Client type: Confidential`
    -  set `Authorization grant type: Authorization code`
    -  set `Skip authorization: True`

1. Upload some geospatial file to geonode. By default GeoNode wants to make all datasets public, but in this case we want to explicitly make it private in order to test out the oauth2 stuff. Modify the uploaded layer's permissions to make sure it is visible only to your current geonode user

1. In QGIS, setup a new WMS provider. 

    -  Set some logical name for the connection
    -  Set `url: http://localhost/geoserver/ows`
    -  Create a new oauth2 configuration and set the following:
        - Request URL: http://localhost/o/authorize/
	- Token URL: http://localhost/o/token/
	- Client ID: the same you got from GeoNode
	- Client secret: the same you got from GeoNode
	- Scope: read write groups

1. Finally, try to use this new WMS connection to load your layer in QGIS. If all goes well, QGIS should redirect you to your browser, where you will receive the error I reported above

1. If you now simply edit the browser's URL bar and modify the request's query part to have `state=` instead of `state` and then reload the request everything shall work OK. The browser shall notify you to go back to QGIS, where your layer will have been recognised and be ready to load.

[geonode's docker-compose]: https://github.com/GeoNode/geonode/blob/master/docker-compose.yml

**QGIS and OS versions**

QGIS version | 3.14.0-Pi | QGIS code revision | 9f7028fd23
-- | -- | -- | --
Compiled against Qt | 5.12.8 | Running against Qt | 5.12.8
Compiled against GDAL/OGR | 3.0.4 | Running against GDAL/OGR | 3.0.4
Compiled against GEOS | 3.8.0-CAPI-1.13.1 | Running against GEOS | 3.8.0-CAPI-1.13.1
Compiled against SQLite | 3.31.1 | Running against SQLite | 3.31.1
PostgreSQL Client Version | 12.2 (Ubuntu 12.2-4) | SpatiaLite Version | 4.3.0a
QWT Version | 6.1.4 | QScintilla2 Version | 2.11.2
Compiled against PROJ | 6.3.1 | Running against PROJ | Rel. 6.3.1, February 10th, 2020
OS Version | Ubuntu 20.04 LTS
Active python plugins | qgis_resource_sharing; pluginbuilder3; Mergin; plugin_reloader; dataset_qa_workbench; geocatbridge; qgis-maptiler-plugin; MetaSearch; db_manager; processing

**Additional context**

**Comment by symbioquine** [comment]
A (hacky) work-around for this bug seems to be putting `state=1` as an extra initial request parameter;

![image](https://user-images.githubusercontent.com/30754460/100547773-7303c780-321d-11eb-9e75-7bbb2f2ad0b1.png)

(At least when used to authorize against [Drupal's oauth2_server module](https://www.drupal.org/project/oauth2_server) - YMMV with other implementations.)

**Comment by rouault** [comment]
This has been fixed in 3.16.0 per https://github.com/qgis/QGIS/pull/39552

**Comment by symbioquine** [comment]
Actually disregard my previous comment. I'm using 3.16.0 and, as @rouault's link would suggest, the `state` parameter was correctly being (completely) omitted. The requests were failing because the `state` parameter is required by the oauth2_server module.
